### PR TITLE
Write database models, HTTP API and tests for backend

### PR DESCRIPTION
**Pull request by iafisher**
We shouldn't have to make too many changes to the backend after this,
except to serve the HTML and JS from the ui app. I may add some more
unit tests for the error handling in the HTTP API.

I wrote the API and models with the full application in mind, not just
the barebones initial implementation, so we won't have to go back and
change things later.

### CODE REVIEW

File: backend/api/admin.py [hunk 1]
```diff
@@ -1,3 +1,13 @@
 from django.contrib import admin
 
-# Register your models here.
+from .models import Project, Task
+
+
+class ProjectAdmin(admin.ModelAdmin):
+    # This tells Django to automatically fill in the `slug` field by slugifying the
+    # `name` field.
+    prepopulated_fields = {"slug": ("name",)}
```
Comment (noahdkim) [inline backend/api/admin.py]: what is slugifying?

File: backend/api/views.py [hunk 2]
```diff
@@ -1,3 +1,219 @@
...
+                types_as_str = " or ".join(v.__name__ for v in value)
+                raise ApiError(f"{key} should have type {types_as_str}")
+        else:
+            if not isinstance(payload[key], value):
+                raise ApiError(f"{key} should have type {value.__name__}")
+
+    return payload
+
+
+def _parse_description(description):
```
Comment (noahdkim) [inline backend/api/views.py]: Is the frontend going to be rendering the short and long descriptions differently?

File: backend/api/admin.py (hunk 1 above)
Comment (iafisher) [reply backend/api/admin.py]: It's the process of converting an arbitrary string into a string that's suitable to be part of a URL. You can read about the details of the Django implementation here: https://docs.djangoproject.com/en/3.0/ref/utils/#django.utils.text.slugify

File: backend/api/views.py (hunk 2 above)
Comment (iafisher) [reply backend/api/views.py]: Yes, the short descriptions will always be visible, while the long descriptions will only be visible when a task is selected.

**Comment by iafisher** [comment]
> Just to check -- this PR doesn't include any code involved in actually serving the JS right? it's just model logic?

Correct

**Pull request merged**